views/course_form/views.py

```diff
@@ -23,8 +61,108 @@ def post(self, request, code: str | None = None, semester: str | None = None):
         Side-effects: New Course, CourseSection, and LabSection models are added to the DB
         TODO: Describe request structure when testing
         '''
-        return render(request, 'login/login.html')
 
+        if not request.user.is_authenticated or request.user.role != "Admin":
+            return redirect("login")
+
+            # Gather form data
+        code = request.POST.get("code", "").strip()
+        name = request.POST.get("name", "").strip()
+        semester = request.POST.get("semester", "").strip()
+        ta_list = request.POST.get("ta_list", "").strip()
+
+        # Validate inputs
+        errors = []
+
+        # Critical validation: Duplicate course code
+        if Course.objects.filter(course_code=code, semester__semester_name=semester).exists():
+            return HttpResponse(repr(['The selected course code already exists for this semester']),
+                                content_type="text/plain", status=200)
+
+        # Validate course code
+        if not code:
+            errors.append('Course code must be a valid value')
+            # use regex to check for invalid characters
+        elif not re.match(r'^[A-Za-z0-9]+$', code):
+            errors.append("Course code must be a valid value")  # Same error message for consistency
+
+            # validate name
+        if not name:
+            errors.append('Course name must be a valid value')
+        elif not re.match(r'^[A-Za-z0-9 ]+$', name):  # Allow alphanumeric and spaces
+            errors.append('Course name must be a valid value')
+
+        # Validate semester
+        if not CourseController.CourseController.validate_semester(semester):
+            return HttpResponse(
+                repr(["Course semester must be a valid semester"]),
+                content_type="text/plain",
+                status=200
+            )
+
+        if not CourseController.CourseController.validate_semester(semester):
+            errors.append('Course semester must be a valid semester')
+
+            # Validate TA list
+        if ta_list:
+            ta_usernames = [ta.strip() for ta in ta_list.split(",")]
+            non_existent_tas = [
+                ta for ta in ta_usernames if not User.objects.filter(username=ta).exists()
+            ]
+            if non_existent_tas:
+                errors.append('One or more of the selected TAs is not an existing user')
+
+        # TA Role validation
+        if ta_list:
+            ta_usernames = [ta.strip() for ta in ta_list.split(",")]
+            invalid_tas = [
+                ta for ta in ta_usernames
+                if not User.objects.filter(username=ta, role="TA").exists()
```

Comment:
Do not access DB from the view.

Refactor the course controller to throw better exceptions if anything. Or reasonably refactor tests.

Reply:
and this is the first MR. It doesn't need to pass all tests so just let them fail for now. The views shouldn't do DB access.